src/hotspot/cpu/aarch64/gc/z/zGlobals_aarch64.cpp

```diff
@@ -135,9 +138,66 @@
 //  * 63-48 Fixed (16-bits, always zero)
 //
 
+// Default value if probing is not implemented for a certain platform: 128TB
+#define DEFAULT_MAX_ADDRESS_BIT 47
+// Minimum value returned, if probing fails: 64GB
+#define MINIMUM_MAX_ADDRESS_BIT 36
+
+static size_t probe_valid_max_address_bit() {
+#ifdef LINUX
+  size_t max_address_bit = 0;
+  const size_t page_size = os::vm_page_size();
+  for (int i = DEFAULT_MAX_ADDRESS_BIT; i > MINIMUM_MAX_ADDRESS_BIT && max_address_bit == 0; --i) {
+    const uintptr_t base_addr = ((uintptr_t) 1U) << i;
+    if (msync((void*)base_addr, page_size, MS_ASYNC) == 0) {
+      // msync suceeded, the address is valid, and maybe even already mapped.
+      max_address_bit = i;
+      break;
+    }
```

Comment:
Is there a one-off error here? Taking i == 47 as an example. This means that you test the base_address == '10000000 00000000 00000000 00000000 00000000 00000000' (in bits). That is, you test that the address range with the 48th bit set (128T-256T) is usable. However, when 47 then is returned to the caller, it interprets it as if the 64T-128T range is usable.

Reply:
Good catch. I verified your assumption by doing the following:
I returned the result of probe_valid_max_address_bit() + 1 from ZPlatformAddressOffsetBits(). I then executed a VM on a platform with only 3 pagetable levels, which has less than DEFAULT_MAX_ADDRESS_BIT bits available in the address space. The heap allocation succeeded, which means the current patch has a one off error.

However, I would fix this in ZPlatformAddressOffsetBits(). because I interpret the result of probe_valid_max_address_bit() as to be the highest bit index of an address which can be allocated. Which means that mmap the result of 1 << probe_valid_max_address_bit() should succeed (if the page is not already mapped). I think ZPlatformAddressOffsetBits() should add 1 to the result of probe_valid_max_address_bit().

Reply:
Sorry, the explanation of my verification was wrong. I did the following:
`
size_t ZPlatformAddressOffsetBits() {
  const static size_t valid_max_address_bit = probe_valid_max_address_bit();
  const size_t max_address_offset_bits = valid_max_address_bit - 3;
  return max_address_offset_bits + 1;
}
`
and allocation of the Java heap succeeded.